mongofiles/mongofiles.go

```diff
@@ -57,185 +62,281 @@ type MongoFiles struct {
 	// filename in GridFS
 	FileName string
 
-	//ID to put into GridFS
+	// ID to put into GridFS
 	Id string
+
+	// GridFS bucket to operate on
+	bucket *gridfs.Bucket
 }
 
-// GFSFile represents a GridFS file.
-type GFSFile struct {
-	Id          bson.ObjectId `bson:"_id"`
-	ChunkSize   int           `bson:"chunkSize"`
-	Name        string        `bson:"filename"`
-	Length      int64         `bson:"length"`
-	Md5         string        `bson:"md5"`
-	UploadDate  time.Time     `bson:"uploadDate"`
-	ContentType string        `bson:"contentType,omitempty"`
+// Struct representing a GridFS files collection document.
+type gfsFile struct {
+	Id         interface{}        `bson:"_id"`
+	Name       string             `bson:"filename"`
+	Length     int64              `bson:"length"`
+	Md5        string             `bson:"md5"`
+	UploadDate time.Time          `bson:"uploadDate"`
+	Metadata   gfsFileMetadata    `bson:"metadata"`
+
+	// Storage required for reading and writing GridFS files
+	mf         *MongoFiles
+	downStream *gridfs.DownloadStream
+	upStream   *gridfs.UploadStream
 }
 
-// ValidateCommand ensures the arguments supplied are valid.
-func (mf *MongoFiles) ValidateCommand(args []string) error {
-	// make sure a command is specified and that we don't have
-	// too many arguments
-	if len(args) == 0 {
-		return fmt.Errorf("no command specified")
-	}
+// Struct representing the metadata associated with a GridFS files collection document.
+type gfsFileMetadata struct {
+	ChunkSize   int           	   `bson:"chunkSize"`
+	ContentType string             `bson:"contentType,omitempty"`
+}
 
-	switch args[0] {
-	case List:
-		if len(args) > 2 {
-			return fmt.Errorf("too many positional arguments")
-		}
-		if len(args) == 1 {
-			mf.FileName = ""
-		} else {
-			mf.FileName = args[1]
-		}
-	case Search, Put, Get, Delete:
-		if len(args) > 2 {
-			return fmt.Errorf("too many positional arguments")
+// Write data to GridFS Upload Stream. If this file has not been written to before, this function will open a new stream that must be closed.
+// Note: if this file already exists, the chunks written here will be orphaned when close is called and an error will be returned.
```

Comment:
so the write call will succeed but close will later fail if the file already exists? that seems a bit unexpected. is this just a limitation of how the GridFS API works?

Reply:
> I just checked to see what the current mongofiles behavior is, and it seems it actually _deletes_ the old chunks after failing to insert into the chunks collection, which is a serious bug.

by the old chunks, do you mean the ones from the file that already exists? 😱 

I don't think it is worth checking ahead of time for duplicates. from the user perspective, the behavior is really the same, it's just that we would potentially identify the problem like three lines earlier in `MongoFiles.put`.

Reply:
is there a test checking behavior when you try to write to an existing file without using the `replace` option? given the bug you described before, I think it would be worth adding one.

Reply:
>by the old chunks, do you mean the ones from the file that already exists? 😱

Yeah. I filed [TOOLS-2229](https://jira.mongodb.com/browse/TOOLS-2229) for that.

>is there a test checking behavior when you try to write to an existing file without using the `replace` option? given the bug you described before, I think it would be worth adding one.

I added one and in doing so discovered why the docstring was written the way it was. The go driver actually has a bug where the `{files_id: 1, _id: 1}` index is not _unique_, as the spec requires. So we actually are getting orphan chunks in this case. Filed [GODRIVER-856](https://jira.mongodb.com/browse/GODRIVER-856) for that one.

I'm going to hold off on merging the test into this patch though, as I'll likely implement it when I fix TOOLS-2229 (which will come before this port is done).

Reply:
all sounds good 👍